### PR TITLE
feat: add more ticks to the histogram

### LINKED ISSUE #42: Adding more ticks to histogram
I've been playing a bit with the histogram and having just the min and max was a bit hard to read it.  I've added more ticks, what do you think?  I would like it if there were some `|` or something showing where the value actually aligns, but even this is good for a rough estimate.

Image

### PR DESCRIPTION
**Pull request by Fuco1**
Fixes #42 

I tried to test a bunch of edge cases but there could be situations where it might break.  I was thinking as part of my learning ocaml I could set up some test cases.  Do you have/is there some prefered testing framework?  I'm kind of a testing nerd tbh :D

### CODE REVIEW

File: src/histogram.ml [hunk 1]
```diff
@@ -93,14 +93,35 @@ let output ?(unit = "") histogram ~w ~h =
         done;
         print_newline ();
         let string_of_float x =
-          if x = 0. || x >= 1. then string_of_int (int_of_float x) else string_of_float x
+          (* if x = 0. || x >= 1. then string_of_int (int_of_float x) else string_of_float x *)
```
Comment (Fuco1) [inline src/histogram.ml:96]: I'm not sure what was the point of this logic, I replaced it only with a simple `printf`, since otherwise it would print `1 div`, `1 div`, `1 div`, instead of `1`, `1.2` and `1.4` for example.

File: src/histogram.ml [hunk 2]
```diff
@@ -93,14 +93,35 @@ let output ?(unit = "") histogram ~w ~h =
...
-        let blank = w - String.length min_string - String.length max_string in
-        print_string min_string;
-        print_string (String.make blank ' ');
-        print_endline max_string;
+        let format_currency x cur = string_of_float x ^ " " ^ cur in
+        let step = (max_value -. min_value) /. 5. in
+        let printed_space = ref 0 in
+        for i = 0 to 5 do
+          let label_len = (String.length @@ format_currency (min_value +. float_of_int i *. step) unit) in
+          printed_space := !printed_space + label_len;
```
Comment (Fuco1) [inline src/histogram.ml:104]: Is the use of refs like this generally frowned upon?  I mean people hate mutability for a reason :D

File: src/histogram.ml [hunk 3]
```diff
@@ -93,14 +93,35 @@ let output ?(unit = "") histogram ~w ~h =
...
+        done;
+        let get_step_and_overflow x =
+          (Float.trunc x, x -. Float.trunc x)
+        in
+        let overflow = ref 0. in
+        for i = 0 to 5 do
+          let (blank, over) = (get_step_and_overflow @@ (float_of_int (w - !printed_space)) /. 5.) in
+          overflow := !overflow +. over;
+          let blank =
+            if !overflow >= 1. then begin
```
Comment (Fuco1) [inline src/histogram.ml:114]: I don't know why this code wouldn't work without wrapping with `begin` `end` which I just googled somewhere.  You haven't used it anywhere so there is probably always a way to write things better.  

Actually I should try that formatter that you mentioned, it might get rid of it.

File: src/histogram.ml (hunk 2 above)
Comment (doomeer) [reply src/histogram.ml:104]: It is frown upon yes… sometimes. Mostly by people who have a background in Haskell I would say :D OCaml is a bit more pragmatic. References do feel dirty but sometimes they are just easier and a good compromise.

The bigger the scope of a reference is, the more dirty it feels. For instance, consider global references. Global = the biggest scope = the most dirty use of references.
- The main issue is that it makes the code not thread-safe. If one day we want to, say, use the new parallelism features from OCaml 5 to speed up the simulation by running multiple ones in parallel, global references would be problematic (only if we actually modify them though).
- In general it also makes it hard to reason about the code because you have to think about all the places that could mutate it.

When a reference is local to a function though, I would argue it's totally fine. Sure you could use a recursive function instead of a `for` loop to avoid a reference, but it would arguably make the code less readable.

If you want you can limit the scope of the reference further by shadowing the variable (with `let printed_space = !printed_space in`) once you know you no longer need to modify it. This effectively limits the scope of the reference and makes it less dirty.

File: src/histogram.ml (hunk 3 above)
Comment (doomeer) [reply src/histogram.ml:114]: `begin ... end` is equivalent to parentheses `(...)`. I prefer parentheses, some people prefer `begin ... end`. In this particular case you do need either of them (*).

(*) There is actually a trick to avoid parentheses. If you write a `let`, the scope of the `let` will take care of parentheses for you. So you could write something like:
```
            if !overflow >= 1. then begin
              let () = () in
              overflow := !overflow -. 1.;
              blank +. 1.
            end else
              blank
```
I wouldn't particularly recommend to introduce a `let` just for that though.

File: src/histogram.ml (hunk 1 above)
Comment (Fuco1) [reply src/histogram.ml:96]: Yea that makes sense, since you set up the scales to go from let's say 1 to 5 and it only printed the edge values.

If now it would print 6 points the interpolated values will be decimals, there's no way around it if the range is small.  In which cases we could also convert to `c` instead of `div`, but honestly sometimes it's easier to just count in divs... it could be added as an option though, to print in a specific currency.

Anyway, I'm not sure which one I prefer `1 div      1.2 div       1.4 div` or `1.0 div     1.2 div       1.4 div`, with the extra `.0` for consistency.  Maybe the second one is slightly better.

If the range gets bigger than the tick logic could also be made better to always pick nice numbers, so even if it goes from 0 to 9 it would print 0, 2, 4, 6, 8, 9 instead of 0, (1/5 * 9)... whatever that fraction would come to be :D

I'm using R a lot and they do plot the axes with nice values, I might check the algorithm from ggplot to see how they do it.
